encoding/int_encode.go

```diff
@@ -1,30 +1,216 @@
+	enc.reset(writer)
+
+	// Determine whether we want to do table compression.
+	var (
+		firstVal int64
+		max      = math.MinInt64
+		min      = math.MaxInt64
+		idx      int
+		dict     = make(map[int]int, dictEncodingMaxCardinalityInt)
+	)
+	for valuesIt.Next() {
+		curr := valuesIt.Current()
+
+		if idx == 0 {
+			firstVal = int64(curr)
+		}
+
+		// Only grow the dict map if we are below dictEncodingMaxCardinalityInt.
+		// This way we track if we've exceeded the max # of uniques.
+		if len(dict) <= dictEncodingMaxCardinalityInt {
+			if _, ok := dict[curr]; !ok {
+				dict[curr] = idx
+				idx++
+			}
+		}
+
+		// Track max/min.
+		if curr < min {
+			min = curr
+		}
+		if curr > max {
+			max = curr
+		}
+	}
+
+	if valuesIt.Err() != nil {
+		return valuesIt.Err()
+	}
+
+	// Rewind iteration.
+	valuesIt.Rewind()
+
+	if len(dict) <= dictEncodingMaxCardinalityInt {
+		// Table encode if there is less than 256 unique valuesIt.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+		// Min value used to calculate dictionary values.
+		//Each dictionary value is a positive number added to the min value.
+		enc.metaProto.MinValue = int64(min)
+		// Min number of bytes to encode each value into the int dict.
+		enc.metaProto.BytesPerDictionaryValue = int64(math.Ceil(float64(bits.Len(uint(max-min))) / 8.0))
+		// Min number of bits required to encode table indices.
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(len(dict))))
+	} else {
+		// Default to delta encoding.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DELTA
+		enc.metaProto.DeltaStart = firstVal
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint((max - min) + 1))) // Add 1 for the sign bit.
+	}
+
+	if err := proto.EncodeIntMeta(&enc.metaProto, &enc.buf, writer); err != nil {
+		return err
+	}
+
+	switch enc.metaProto.Encoding {
+	case encodingpb.EncodingType_DICTIONARY:
+		if err := enc.encodeDictionary(writer, enc.metaProto.BytesPerDictionaryValue, enc.metaProto.BitsPerEncodedValue, dict, valuesIt); err != nil {
+			return err
+		}
+	case encodingpb.EncodingType_DELTA:
+		if err := enc.encodeDelta(enc.metaProto.BitsPerEncodedValue, valuesIt); err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 // Reset resets the encoder.
-func (enc *IntEnc) Reset() { panic("not implemented") }
+func (enc *IntEnc) reset(writer io.Writer) {
+	// Reset the bitWriter at the start of every `Encode` call.
+	enc.bitWriter.Reset(writer)
+	enc.metaProto.Reset()
+	enc.dictionaryProto.Reset()
+}
+
+func (enc *IntEnc) encodeDelta(
+	bitsPerEncodedValue int64,
+	valuesIt RewindableIntIterator,
+) error {
+	// Encode the first value which is always a delta of 0.
+	if !valuesIt.Next() {
+		return valuesIt.Err()
+	}
+
+	// Write an extra bit to encode the sign of the delta.
+	if err := enc.bitWriter.WriteBits(uint64(0), int(bitsPerEncodedValue)); err != nil {
+		return err
+	}
+
+	negativeBit := 1 << uint(bitsPerEncodedValue)
```

Comment:
Hm should this be `bitsPerEncodedValue - 1` now? Guess we don't have a unit test to catch this?

Reply:
Adding unit tests for negative numbers too.